staging/sockets/luasockets.c

```diff
@@ -0,0 +1,205 @@
+	else
+		luaL_error(L, "'%s' domain not supported\n", domain);
+
+	if (strcmp(type, "dgram") == 0 || strcmp(type, "datagram") == 0)
+		typ = SOCK_DGRAM;
+	else if (strcmp(type, "stream"))
+		typ = SOCK_STREAM;
+	else
+		luaL_error(L, "'%s' type not supported\n", type);
+
+	if ((err = fsocreate(dom, &so, typ, 0, &sfd)))
+		luaL_error(L, "could not create socket (error %d)", err);
+
+	luasocket = (struct luasocket*)
+		lua_newuserdata(L, sizeof(struct luasocket));
+	luasocket->socket = so;
+	luasocket->fd = sfd;
+
+	return 1;
+}
+
+static int
+sockets_close(lua_State *L)
+{
+	struct luasocket *luasocket;
+	int err;
+
+	if ((luasocket = (struct luasocket *) lua_touserdata(L, 1)) == NULL)
+		luaL_error(L, "invalid Lua socket\n");
+	if ((err = soclose(luasocket->socket)) != 0)
+		luaL_error(L, "could not close socket (error %d)\n", err);
+	lua_pushboolean(L, 1);
+	return 1;
+}
+
+typedef int BindOrConnect(struct socket*, struct sockaddr*, struct lwp*);
+
+static int
+sockets_bind_connect(BindOrConnect handler, lua_State *L)
+{
+	struct luasocket *luasocket;
+	const char *domn;
+	const char *addr;
+	int err;
+
+	if ((luasocket = (struct luasocket *) lua_touserdata(L, 1)) == NULL)
+		luaL_error(L, "invalid socket");
+	domn = luaL_checkstring(L, 2);
+	addr = luaL_checkstring(L, 3);
+
+	// XXX inet4 and inet6 could share code
+	// XXX we could also declare a sockaddr in the function scope
+	// and then specialize it to sockaddr_in(,6, un) in the branches below
+	if (strcmp(domn, "inet4") == 0) {
+		struct sockaddr_in in4addr;
+
+		if (lua_gettop(L) != 4)
+			luaL_error(L, "inet4 expects an IP address and a port number");
+
+		memset(&in4addr, 0, sizeof(in4addr));
+		in4addr.sin_len = sizeof(struct sockaddr_in);
+		in4addr.sin_addr.s_addr = inet_addr(addr);
+		in4addr.sin_port = htons(luaL_checkinteger(L, -1));
+		in4addr.sin_family = AF_INET;
+
+		solock(luasocket->socket);
+		if ((err = handler(luasocket->socket, (struct sockaddr*) &in4addr,
+			curlwp)) != 0)
+			luaL_error(L, "could not connect");
+		sounlock(luasocket->socket);
+
+		lua_pushboolean(L, 1);
+	} else if (strcmp(domn, "local") == 0) {
+		;
+	}
+
+	return 1;
+}
+
+static int
+sockets_connect(lua_State *L)
+{
+	return sockets_bind_connect(soconnect, L);
+}
+
+static int
+sockets_bind(lua_State *L)
+{
+	return sockets_bind_connect(sobind, L);
+}
+
+static int
+sockets_write(lua_State *L)
+{
+	struct luasocket* luasocket;
+	const char *buff;
+	int cnt;
+
+	if ((luasocket = (struct luasocket*) lua_touserdata(L, 1)) == NULL)
+		luaL_error(L, "invalid socket");
+	buff = luaL_checkstring(L, 2);
+
+	//XXX handle binary data
+	cnt = write(luasocket->fd, buff, strlen(buff));
+	lua_pushinteger(L, cnt);
+
+	return 1;
+}
+
+static int
+sockets_read(lua_State *L)
+{
+	struct luasocket* luasocket;
+	//XXX probably not a good idea; lua_pushlstring?
+	char buff[BUF_SIZE];
+	int cnt;
+
+	if ((luasocket = (struct luasocket*) lua_touserdata(L, 1)) == NULL)
+		luaL_error(L, "invalid socket");
```

Comment:
you should also receive what you want to read, such as http://www.lua.org/manual/5.3/manual.html#pdf-file:read